PLAT-10563: FormReply activity

symphony-youness opened the pull request:
### Ticket
[PLAT-10563](https://perzoinc.atlassian.net/browse/PLAT-10563)

### Description
Base classes for a FormReply Activity.

Added a quick fix in the ActivityRegistry: The problem was that the `act.before_matcher()` method was called before checking the correct instance of the activity. 

Documentation is coming ...

### Checklist
- [x] Referenced a ticket in the PR title and in the corresponding section
- [x] Filled properly the description and dependencies, if any
- [x] Unit tests updated or added
- [x] Docstrings added or updated
- [x] Updated the documentation in [docs folder](../docs)

symphony-youness committed:
PLAT-10563 Implementation of Form Reply activity
- Implementation
- fixed small error in the activity registry: checking for the correct activity instance before calling the before_matcher method.
- moved activities example to new folder
- added resource package for examples

symphony-youness committed:
PLAT-10563 Added tests
- Tests for FormReply and Command activity
- Handling of MessageParserError in a Command activity when the text in incorrect PresentationML format

symphony-youness committed:
PLAT-10563 Updated api and registry docstrings

symphony-youness committed:
PLAT-10563 Added registry test
on_symphony_elements_action test

symphony-youness committed:
PLAT-10563 Added documentation

symphony-youness commented on docsrc/markdown/activity-api.md:
````diff
@@ -75,3 +75,70 @@ try:
 ```
+1. The `matches()` method allows the activity logic to be triggered when a message starts by a mention to the bot and the text `/hello` separated by a space. Ex: `@bot_name /hello`
+2. The activity logic. Here, we send a message: "Hello, World"
+## Form Activity
````
Will add a break on top of this.

symphony-elias commented on examples/activities/activity.py:
```diff
@@ -27,8 +26,7 @@ async def on_activity(self, context: CommandContext):
 
-logging.config.fileConfig(os.path.dirname(os.path.abspath(__file__)) + '/logging.conf',
-                          disable_existing_loggers=False)
+logging.config.fileConfig(Path("../logging.conf"), disable_existing_loggers=False)
```
Beware, I am not sure this will work if you run the script from somewhere else than the script folder.

symphony-youness replied:
Indeed, good catch. Will fix

symphony-elias commented on symphony/bdk/core/activity/form.py:
```diff
@@ -0,0 +1,50 @@
+        pass
+
+    def before_matcher(self, context: FormReplyContext):
+        context._form_id = context.source_event.form_id
```
I'd prefer `context._form_id` and `context._form_values` to be set in the constructor of `FormReplyContext` directly

symphony-youri replied:
as the event is coming from the outside should we be defensive about what we get? (like if it is null or does not have a form id)

symphony-youri replied:
> I'd prefer `context._form_id` and `context._form_values` to be set in the constructor of `FormReplyContext` directly

true, I wonder why we have the before_matcher() if we can do part of the context's constructor

symphony-youness replied:
I tried sending a form without an `id` and It didn't work: ` {"code":400,"message":"Failed to parse MessageML: The attribute \"id\" is required"} so is that possible ?

I thought it was to separate the logic of modifying the context in `before_matcher()` before validating it in `matches()`

To handle the possible unwanted values we get, maybe we can do this, WDT @symphony-elias :
```Python
self._form_id = getattr(source_event, "form_id", None)
or
self._form_id = getattr(source_event, "form_id", "")
```
We definitely can add those to the constructor. Maybe I am missing the purpose of thibault setting them in `before_matcher()`

symphony-youri commented on examples/activities/form_activity.py:
```diff
@@ -0,0 +1,56 @@
+class SlashGifCommandActivity(CommandActivity):
+    def __init__(self, messages: MessageService):
+        self._messages = messages
+        super().__init__()
```
why do we call init here and not in the ReplyFormReplyActivity?

symphony-youness replied:
**I will remove this along with the constructor of the `CommandActivity`**

```Python
def __init__(self):
    self.bot_display_name = str
```
This was done in java and I think it's purpose was to set the bot name in the activity.
ActivityRegistery.java : `a command activity (potentially) needs the bot display name in order to parse the message text content` (maybe we don't actually need it)

symphony-youri commented on tests/core/activity/registry_test.py:
```diff
@@ -62,3 +73,16 @@ async def test_on_message_sent(activity_registry, message_sent, command):
+
+
+@pytest.mark.asyncio
+async def test_on_symphony_elements_action(activity_registry, elements_action, form):
```
maybe we could add tests to cover different cases like the event not being an ElementsAction
and the matcher not returning true
(right now we have 100% lines coverage but not 100% branch coverage)

symphony-youness replied:
will do

symphony-youness committed:
PLAT-10563 Setting context fields in the constructor
- Set context fields in the constructor
- Updated documentation to show the use of the `before_matcher()` method

symphony-youness committed:
PLAT-10563 Fixed and added meaningful tests
- registry tests
- updated command tests to match the changes in the activities API
